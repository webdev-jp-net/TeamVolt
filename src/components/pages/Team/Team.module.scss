@use "style/_variable" as *;

.team {
  display: grid;
  grid-template-rows: auto 1fr auto;
  height: 100%;
}

// ヘッダー
.header {
  display: flex;
  flex-direction: column;
  gap: rem(24);
  margin: rem(24) rem(24) auto;
}

// リスト
.list {
  display: flex;
  flex-direction: column;
  gap: rem(12);
  padding: 0;
  margin: rem(24);
}

.item {
  padding: rem(12) 0;
  margin: 0;
  list-style: none;
  border-bottom: 1px solid #000;

}


// 追加フォーム
.addForm {
  display: flex;
  gap: rem(12);
  margin: rem(24);
}

.input {
  width: 100%;
}


// フッター
.footer {
  display: flex;
  flex-direction: column;
  gap: rem(24);
  margin: auto rem(24) rem(24);
}

