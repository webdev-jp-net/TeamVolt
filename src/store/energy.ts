import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { createApi } from '@reduxjs/toolkit/query/react';
import { doc, getDoc, updateDoc, arrayUnion, runTransaction } from 'firebase/firestore';
import { db } from 'firebaseDB';

type ChargeUnitsData = { member: string; count: number };

type State = {
  genEnergy: number;
  chargeUnits: ChargeUnitsData[];
};
const initialState: State = {
  genEnergy: 0,
  chargeUnits: [],
};

// RTK Queryの設定
// https://redux-toolkit.js.org/rtk-query/overview

export const energyGetApi = createApi({
  reducerPath: 'energyGetApi',
  baseQuery: async id => {
    if (id) {
      const docRef = doc(db, 'team', id);
      const docSnap = await getDoc(docRef);
      if (docSnap.exists()) {
        return { data: docSnap.data().chargeUnits };
      } else {
        return { error: 'No such document!' };
      }
    } else {
      return { error: 'No such document!' };
    }
  },
  endpoints: builder => ({
    // 獲得バッテリー取得
    getChargeUnits: builder.query<ChargeUnitsData[], string>({
      query: id => id,
    }),
  }),
});
export const { useGetChargeUnitsQuery } = energyGetApi;

export const energyPostApi = createApi({
  reducerPath: 'energyPostApi',
  baseQuery: async ({ id, value }: { id?: string; value: ChargeUnitsData }) => {
    if (id && value) {
      const docRef = doc(db, 'team', id);
      try {
        await runTransaction(db, async transaction => {
          const sfDoc = await transaction.get(docRef);
          if (!sfDoc.exists()) {
            // eslint-disable-next-line no-throw-literal
            throw 'Document does not exist!';
          }
          const oldList = sfDoc.data().chargeUnits || [];
          const newList = [
            ...oldList.filter((item: ChargeUnitsData) => item.member !== value.member),
            value,
          ];
          transaction.update(docRef, { chargeUnits: newList });
        });
      } catch (e) {
        console.log('Transaction failure:', e);
      }
      // レスポンスで返す情報を再取得
      const docSnap = await getDoc(docRef);
      if (docSnap.exists()) {
        return { data: docSnap.data().chargeUnits };
      } else {
        return { error: 'No such document!' };
      }
    } else {
      return { error: 'No such document!' };
    }
  },
  endpoints: builder => ({
    // 獲得バッテリー追加
    addChargeUnits: builder.mutation<ChargeUnitsData[], { id: string; value: ChargeUnitsData }>({
      query: ({ id, value }) => ({
        operationType: 'remove_challenger',
        id,
        value,
      }),
    }),
  }),
});

export const { useAddChargeUnitsMutation } = energyPostApi;

const energy = createSlice({
  name: 'energy',

  initialState,

  reducers: {
    // 獲得エネルギーを更新
    updateGenEnergy: (state, action: PayloadAction<number>) => {
      return {
        ...state,
        genEnergy: action.payload,
      };
    },
    // 獲得エネルギーをリセット
    resetEnergy: () => {
      return {
        genEnergy: 0,
        chargeUnits: [],
      };
    },
  },

  extraReducers: builder => {
    // 成功: 獲得バッテリー取得
    builder.addMatcher(
      energyGetApi.endpoints.getChargeUnits.matchFulfilled,
      (state, action: PayloadAction<ChargeUnitsData[]>) => {
        const result = { chargeUnits: action.payload };
        return {
          ...state,
          ...result,
        };
      }
    );
    // 成功: 獲得バッテリー追加
    builder.addMatcher(
      energyPostApi.endpoints.addChargeUnits.matchFulfilled,
      (state, action: PayloadAction<ChargeUnitsData[]>) => {
        const result = { chargeUnits: action.payload };
        return {
          ...state,
          ...result,
        };
      }
    );
  },
});

// Action Creator
export const { updateGenEnergy, resetEnergy } = energy.actions;

// Reducer
export default energy.reducer;
