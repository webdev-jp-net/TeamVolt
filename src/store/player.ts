import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { createApi } from '@reduxjs/toolkit/query/react';
import {
  doc,
  getDoc,
  getDocs,
  updateDoc,
  arrayRemove,
  arrayUnion,
  addDoc,
  deleteField,
  collection,
  runTransaction,
} from 'firebase/firestore';
import { db } from 'firebaseDB';

import type { PlayerArticleData } from 'types/player';
import type { TeamArticleData, ChargeUnitsData, MemberData } from 'types/team';

type State = {
  localId: string;
  player?: string;
  playerList: PlayerArticleData[];
  selectedTeam?: string;
  myTeam?: TeamArticleData;
  genEnergy: number;
};

const initialState: State = {
  localId: '',
  playerList: [],
  genEnergy: 0,
};

// RTK Queryの設定
// https://redux-toolkit.js.org/rtk-query/overview
// 読込用
export const playerGetApi = createApi({
  reducerPath: 'playerGetApi',
  baseQuery: async () => {
    const playerRef = collection(db, 'player');
    const playerSnapshot = await getDocs(playerRef);
    const result = playerSnapshot.docs.map(doc => {
      return { id: doc.id, ...doc.data() };
    });
    return { data: result };
  },
  endpoints: builder => ({
    // プレイヤー情報取得
    getPlayer: builder.query<PlayerArticleData[], void>({
      query: () => ({}),
    }),
  }),
});
export const { useGetPlayerQuery } = playerGetApi;

// チーム情報取得
export const teamArticleGetApi = createApi({
  reducerPath: 'teamArticleGetApi',
  baseQuery: async ({ id }) => {
    const docRef = doc(db, 'team', id);
    const docSnap = await getDoc(docRef);
    if (docSnap.exists()) {
      return { data: { id, ...docSnap.data() } };
    } else {
      return { error: 'No such document!' };
    }
  },
  endpoints: builder => ({
    // チーム情報取得
    getTeamArticle: builder.query<TeamArticleData, string>({
      query: id => ({ id }),
    }),
  }),
});
export const { useGetTeamArticleQuery } = teamArticleGetApi;

// 書込用
export const playerAddApi = createApi({
  reducerPath: 'playerAddApi',
  baseQuery: async ({ value }) => {
    const playerRef = collection(db, 'player');
    const docRef = await addDoc(playerRef, {
      localId: value,
    });
    return { data: { id: docRef.id, localId: value } };
  },
  endpoints: builder => ({
    // プレイヤー情報追加
    addPlayer: builder.mutation<PlayerArticleData, string>({
      query: value => ({ value }),
    }),
  }),
});
export const { useAddPlayerMutation } = playerAddApi;

// チーム情報の更新

// チームメンバー編集
type teamPostApiProps = {
  operationType: string;
  id?: string;
  value?: string | number | ChargeUnitsData | MemberData;
};
export const teamPostApi = createApi({
  reducerPath: 'teamPostApi',
  baseQuery: async ({ operationType, id, value }: teamPostApiProps) => {
    if (operationType === 'add_member' && id) {
      // メンバーを追加
      const docRef = doc(db, 'team', id);
      await updateDoc(docRef, {
        member: arrayUnion(value),
      });
      // レスポンスで返す情報を再取得
      const docSnap = await getDoc(docRef);
      if (docSnap.exists()) {
        return { data: { id, ...docSnap.data() } };
      } else {
        return { error: 'No such document!' };
      }
    } else if (operationType === 'remove_member' && id) {
      const docRef = doc(db, 'team', id);
      // メンバーを削除
      await updateDoc(docRef, {
        member: arrayRemove(value),
      });
      // レスポンスで返す情報を再取得
      const docSnap = await getDoc(docRef);
      if (docSnap.exists()) {
        return { data: { id, ...docSnap.data() } };
      } else {
        return { error: 'No such document!' };
      }
    } else if (operationType === 'add_challenger' && id) {
      const docRef = doc(db, 'team', id);
      // 代表者を登録
      await updateDoc(docRef, {
        challenger: value,
      });
      // レスポンスで返す情報を再取得
      const docSnap = await getDoc(docRef);
      if (docSnap.exists()) {
        return { data: { id, ...docSnap.data() } };
      } else {
        return { error: 'No such document!' };
      }
    } else if (operationType === 'remove_challenger' && id) {
      const docRef = doc(db, 'team', id);
      // 代表者を削除
      await updateDoc(docRef, {
        challenger: deleteField(),
      });
      // レスポンスで返す情報を再取得
      const docSnap = await getDoc(docRef);
      if (docSnap.exists()) {
        return { data: { id, ...docSnap.data() } };
      } else {
        return { error: 'No such document!' };
      }
    } else if (operationType === 'add_charge_units' && id && value) {
      // 獲得バッテリーを登録
      const docRef = doc(db, 'team', id);
      try {
        await runTransaction(db, async transaction => {
          const sfDoc = await transaction.get(docRef);
          if (!sfDoc.exists()) {
            // eslint-disable-next-line no-throw-literal
            throw 'Document does not exist!';
          }
          const oldList = sfDoc.data().chargeUnits || [];
          const newList = [
            ...oldList.filter(
              (item: ChargeUnitsData) =>
                item.member !== (value as unknown as ChargeUnitsData).member
            ),
            value,
          ];
          transaction.update(docRef, { chargeUnits: newList });
        });
      } catch (e) {
        console.log('Transaction failure:', e);
      }
      // レスポンスで返す情報を再取得
      const docSnap = await getDoc(docRef);
      if (docSnap.exists()) {
        return { data: { id, ...docSnap.data() } };
      } else {
        return { error: 'No such document!' };
      }
    } else if (operationType === 'update_used_units' && id) {
      const docRef = doc(db, 'team', id);
      // 救出ミッション試行回数を登録
      await updateDoc(docRef, {
        usedUnits: value,
      });
      // レスポンスで返す情報を再取得
      const docSnap = await getDoc(docRef);
      if (docSnap.exists()) {
        return { data: { id, ...docSnap.data() } };
      } else {
        return { error: 'No such document!' };
      }
    } else if (operationType === 'update_current_positions' && id) {
      const docRef = doc(db, 'team', id);
      // 救出ミッション進捗を登録
      await updateDoc(docRef, {
        currentPosition: value,
      });
      // レスポンスで返す情報を再取得
      const docSnap = await getDoc(docRef);
      if (docSnap.exists()) {
        return { data: { id, ...docSnap.data() } };
      } else {
        return { error: 'No such document!' };
      }
    } else if (operationType === 'close_mission' && id) {
      const docRef = doc(db, 'team', id);
      // メンバー、役割、プレイ結果を削除
      await updateDoc(docRef, {
        member: deleteField(),
        challenger: deleteField(),
        chargeUnits: deleteField(),
        currentPosition: deleteField(),
        usedUnits: deleteField(),
      });
      // レスポンスで返す情報を再取得
      const docSnap = await getDoc(docRef);
      if (docSnap.exists()) {
        return { data: { id, ...docSnap.data() } };
      } else {
        return { error: 'No such document!' };
      }
    } else {
      return { error: 'No such document!' };
    }
  },
  endpoints: builder => ({
    // メンバー追加
    addMember: builder.mutation<TeamArticleData, { id: string; value: MemberData }>({
      query: ({ id, value }) => ({
        operationType: 'add_member',
        id,
        value,
      }),
    }),
    // メンバー削除
    removeMember: builder.mutation<TeamArticleData, { id: string; value: MemberData }>({
      query: ({ id, value }) => ({
        operationType: 'remove_member',
        id,
        value,
      }),
    }),
    // 抽選結果追加
    addChallenger: builder.mutation<TeamArticleData, { id: string; value: string }>({
      query: ({ id, value }) => ({
        operationType: 'add_challenger',
        id,
        value,
      }),
    }),
    // 抽選結果追加
    removeChallenger: builder.mutation<TeamArticleData, { id: string; value: string }>({
      query: ({ id, value }) => ({
        operationType: 'remove_challenger',
        id,
        value,
      }),
    }),
    // 獲得バッテリー追加
    addChargeUnits: builder.mutation<TeamArticleData, { id: string; value: ChargeUnitsData }>({
      query: ({ id, value }) => ({
        operationType: 'add_charge_units',
        id,
        value,
      }),
    }),
    // 救出ミッション試行数を更新
    updateUsedUnits: builder.mutation<TeamArticleData, { id: string; value: number }>({
      query: ({ id, value }) => ({
        operationType: 'update_used_units',
        id,
        value,
      }),
    }),
    // 救出ミッション進捗を更新
    updateCurrentPosition: builder.mutation<TeamArticleData, { id: string; value: number }>({
      query: ({ id, value }) => ({
        operationType: 'update_current_positions',
        id,
        value,
      }),
    }),
    // ミッション情報をクリア
    closeMission: builder.mutation<TeamArticleData, string>({
      query: id => ({
        operationType: 'close_mission',
        id,
      }),
    }),
  }),
});
export const {
  useAddMemberMutation,
  useRemoveMemberMutation,
  useAddChallengerMutation,
  useRemoveChallengerMutation,
  useAddChargeUnitsMutation,
  useUpdateUsedUnitsMutation,
  useUpdateCurrentPositionMutation,
  useCloseMissionMutation,
} = teamPostApi;

const player = createSlice({
  name: 'player',

  initialState,

  reducers: {
    // ローカルID更新
    updateLocalId: (state, action: PayloadAction<string>) => {
      return {
        ...state,
        localId: action.payload,
      };
    },
    // 所属チーム更新
    updateTeam: (state, action: PayloadAction<string>) => {
      return {
        ...state,
        selectedTeam: action.payload,
      };
    },
    // 所属チーム脱退
    escapeTeam: state => {
      return {
        ...state,
        selectedTeam: undefined,
      };
    },
    // 獲得エネルギーを更新
    updateGenEnergy: (state, action: PayloadAction<number>) => {
      return {
        ...state,
        genEnergy: action.payload,
      };
    },
    // 獲得エネルギーをリセット
    resetEnergy: store => {
      return {
        ...store,
        genEnergy: 0,
      };
    },
  },
  extraReducers: builder => {
    // 成功: プレイヤー情報取得
    builder.addMatcher(
      playerGetApi.endpoints.getPlayer.matchFulfilled,
      (state, action: PayloadAction<PlayerArticleData[]>) => {
        state.playerList = action.payload;
      }
    );
    // 成功: プレイヤー追加
    builder.addMatcher(
      playerAddApi.endpoints.addPlayer.matchFulfilled,
      (state, action: PayloadAction<PlayerArticleData>) => {
        const result = { playerList: [...state.playerList, action.payload] };
        return {
          ...state,
          ...result,
        };
      }
    );
    // 成功: チーム情報取得
    builder.addMatcher(
      teamArticleGetApi.endpoints.getTeamArticle.matchFulfilled,
      (state, action: PayloadAction<TeamArticleData>) => {
        state.myTeam = action.payload;
      }
    );
    // 成功: メンバー追加
    builder.addMatcher(
      teamPostApi.endpoints.addMember.matchFulfilled,
      (state, action: PayloadAction<TeamArticleData>) => {
        state.myTeam = action.payload;
      }
    );
    // 成功: メンバー削除
    builder.addMatcher(
      teamPostApi.endpoints.removeMember.matchFulfilled,
      (state, action: PayloadAction<TeamArticleData>) => {
        state.myTeam = action.payload;
      }
    );
    // 成功: 代表者追加
    builder.addMatcher(
      teamPostApi.endpoints.addChallenger.matchFulfilled,
      (state, action: PayloadAction<TeamArticleData>) => {
        state.myTeam = action.payload;
      }
    );
    // 成功: 代表者削除
    builder.addMatcher(
      teamPostApi.endpoints.removeChallenger.matchFulfilled,
      (state, action: PayloadAction<TeamArticleData>) => {
        state.myTeam = action.payload;
      }
    );
    // 成功: 獲得バッテリー追加
    builder.addMatcher(
      teamPostApi.endpoints.addChargeUnits.matchFulfilled,
      (state, action: PayloadAction<TeamArticleData>) => {
        state.myTeam = action.payload;
      }
    );
    // 成功: 救出ミッション試行数更新
    builder.addMatcher(
      teamPostApi.endpoints.updateUsedUnits.matchFulfilled,
      (state, action: PayloadAction<TeamArticleData>) => {
        state.myTeam = action.payload;
      }
    );
    // 成功: 救出ミッション進捗更新
    builder.addMatcher(
      teamPostApi.endpoints.updateCurrentPosition.matchFulfilled,
      (state, action: PayloadAction<TeamArticleData>) => {
        state.myTeam = action.payload;
      }
    );
    // 成功: ミッション情報クリア成功
    builder.addMatcher(
      teamPostApi.endpoints.closeMission.matchFulfilled,
      (state, action: PayloadAction<TeamArticleData>) => {
        state.myTeam = action.payload;
      }
    );
  },
});

// Action Creator
export const { updateLocalId, updateTeam, escapeTeam, updateGenEnergy, resetEnergy } =
  player.actions;

// Reducer
export default player.reducer;
