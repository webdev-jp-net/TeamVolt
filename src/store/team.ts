import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { createApi } from '@reduxjs/toolkit/query/react';
import { getDocs, addDoc, collection } from 'firebase/firestore';
import { db } from 'firebaseDB';
import { TeamArticleData } from 'types/team';

type State = {
  teamList: TeamArticleData[];
};

const initialState: State = {
  teamList: [],
};

// RTK Queryの設定
// https://redux-toolkit.js.org/rtk-query/overview
// read
export const teamGetApi = createApi({
  reducerPath: 'teamGetApi',
  baseQuery: async ({ path }) => {
    const teamRef = collection(db, path);
    const teamSnapshot = await getDocs(teamRef);
    const result = teamSnapshot.docs.map(doc => {
      return { id: doc.id, ...doc.data() };
    });
    return { data: result };
  },
  endpoints: builder => ({
    // チーム情報取得
    getTeam: builder.query<TeamArticleData[], void>({
      query: () => ({ path: 'team' }),
    }),
  }),
});
export const { useGetTeamQuery } = teamGetApi;

// write
export const teamAddApi = createApi({
  reducerPath: 'teamAddApi',
  baseQuery: async ({ value }) => {
    const teamRef = collection(db, 'team');
    const docRef = await addDoc(teamRef, {
      name: value,
      member: [],
    });
    return { data: { id: docRef.id, name: value, member: [] } };
  },
  endpoints: builder => ({
    // チーム情報追加
    addTeam: builder.mutation<TeamArticleData, string>({
      query: value => ({ value }),
    }),
  }),
});
export const { useAddTeamMutation } = teamAddApi;

const team = createSlice({
  name: 'team',

  initialState,

  reducers: {
    updateTeamList: (state, action: PayloadAction<TeamArticleData[]>) => {
      return {
        ...state,
        teamList: action.payload,
      };
    },
  },

  extraReducers: builder => {
    // 成功: チーム情報取得
    builder.addMatcher(
      teamGetApi.endpoints.getTeam.matchFulfilled,
      (state, action: PayloadAction<TeamArticleData[]>) => {
        state.teamList = action.payload;
      }
    );
    // 成功: チーム追加
    builder.addMatcher(
      teamAddApi.endpoints.addTeam.matchFulfilled,
      (state, action: PayloadAction<TeamArticleData>) => {
        const result = { teamList: [...state.teamList, action.payload] };
        return {
          ...state,
          ...result,
        };
      }
    );
  },
});

// Action Creator
export const { updateTeamList } = team.actions;

// Reducer
export default team.reducer;
