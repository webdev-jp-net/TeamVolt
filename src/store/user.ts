import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { createApi } from '@reduxjs/toolkit/query/react';
import {
  doc,
  getDoc,
  setDoc,
  collection,
  runTransaction,
  updateDoc,
  // arrayUnion,
  arrayRemove,
} from 'firebase/firestore';
import { db } from 'firebaseDB';

type State = {
  userId: string;
  memberList: string[];
  drawResult?: string;
};

const initialState: State = {
  userId: '',
  memberList: [],
};

type EntriesResponse = {
  memberList: string[];
};

type DrawResultResponse = {
  userId?: string;
};

// RTK Queryの設定
// https://redux-toolkit.js.org/rtk-query/overview
// 読み込み用のAPI
type FirebaseBaseQueryParams = {
  path: string;
};
const firebaseBaseQuery = async ({ path }: FirebaseBaseQueryParams) => {
  const docRef = doc(db, 'status', path);
  const docSnap = await getDoc(docRef);
  if (docSnap.exists()) {
    return { data: docSnap.data() };
  } else {
    return { error: 'No such document!' };
  }
};
export const statusGetApi = createApi({
  reducerPath: 'statusGetApi',
  baseQuery: firebaseBaseQuery,
  endpoints: builder => ({
    getEntries: builder.query<EntriesResponse, void>({
      query: () => ({ path: 'room_entries' }),
    }),
    getDrawResult: builder.query<DrawResultResponse, void>({
      query: () => ({ path: 'draw_result' }),
    }),
  }),
});
export const { useGetEntriesQuery, useGetDrawResultQuery } = statusGetApi;

// 書き込み用のAPI
type FirebaseMutationParams<T> = {
  path: string;
  value: EntriesResponse | DrawResultResponse | string;
};
const firebaseMutationBaseQuery = async <T>({ path, value }: FirebaseMutationParams<T>) => {
  const statusRef = collection(db, 'status');
  const memberListRef = doc(db, 'status', 'room_entries');

  if (path === 'room_entries_add') {
    // 要素を追加
    // await updateDoc(memberListRef, {
    //   memberList: arrayUnion(value),
    // });
    try {
      await runTransaction(db, async transaction => {
        const sfDoc = await transaction.get(memberListRef);
        if (!sfDoc.exists()) {
          // eslint-disable-next-line no-throw-literal
          throw 'Document does not exist!';
        }
        const oldMemberList = sfDoc.data().memberList || [];
        const newMemberList = [...oldMemberList, value];
        const uniqueMemberList = Array.from(new Set(newMemberList));
        transaction.update(memberListRef, { memberList: uniqueMemberList });
      });
      console.log('Transaction successfully committed!');
    } catch (e) {
      console.log('Transaction failed: ', e);
    }
    return { data: null };
  } else if (path === 'room_entries_remove') {
    // 要素を削除
    await updateDoc(memberListRef, {
      memberList: arrayRemove(value),
    });
    return { data: null };
  } else {
    await setDoc(doc(statusRef, path), value as DrawResultResponse);
    return { data: null };
  }
};
export const statusUpdateApi = createApi({
  reducerPath: 'statusUpdateApi',
  baseQuery: firebaseMutationBaseQuery,
  endpoints: builder => ({
    // updateEntries: builder.mutation<void, string[]>({
    //   query: newMemberList => ({
    //     path: 'room_entries',
    //     value: { memberList: newMemberList },
    //   }),
    // }),
    addEntries: builder.mutation<void, string>({
      query: newMember => ({
        path: 'room_entries_add',
        value: newMember,
      }),
    }),
    removeEntries: builder.mutation<void, string>({
      query: newMember => ({
        path: 'room_entries_remove',
        value: newMember,
      }),
    }),
    updateDrawResult: builder.mutation<void, string>({
      query: newUserId => ({ path: 'draw_result', value: { userId: newUserId } }),
    }),
  }),
});

export const { useAddEntriesMutation, useRemoveEntriesMutation, useUpdateDrawResultMutation } =
  statusUpdateApi;

const user = createSlice({
  name: 'user',

  initialState,

  reducers: {
    updateUserId: (state, action: PayloadAction<string>) => {
      state.userId = action.payload;
    },
    updateMemberList: (state, action: PayloadAction<string[]>) => {
      state.memberList = action.payload;
    },
    resetDrawResult: state => {
      state.drawResult = '';
    },
  },

  extraReducers: builder => {
    builder.addMatcher(
      statusGetApi.endpoints.getEntries.matchFulfilled,
      (state, action: PayloadAction<EntriesResponse>) => {
        const oldMemberList = action.payload.memberList || [];
        const newMemberList = [...oldMemberList, state.userId];
        const uniqueMemberList = Array.from(new Set(newMemberList));

        state.memberList = uniqueMemberList;
      }
    );
    builder.addMatcher(
      statusGetApi.endpoints.getDrawResult.matchFulfilled,
      (state, action: PayloadAction<DrawResultResponse>) => {
        state.drawResult = action.payload.userId;
      }
    );
  },
});

// Action Creators
export const { updateUserId, updateMemberList, resetDrawResult } = user.actions;

// Reducer
export default user.reducer;
