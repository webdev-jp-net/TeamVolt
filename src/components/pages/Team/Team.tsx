import { FC, useCallback } from 'react';

import { useDispatch, useSelector } from 'react-redux';

import { useNavigate } from 'react-router-dom';

import { Button } from 'components/parts/Button';
import { RootState } from 'store';
import { updateTeam } from 'store/player';
import { useGetTeamQuery } from 'store/team';

import styles from './Team.module.scss';

import { AddListItem } from './components/AddListItem';
import { ListItem } from './components/ListItem';

export const Team: FC = () => {
  const dispatch = useDispatch();
  const navigate = useNavigate();

  const { team } = useSelector((state: RootState) => state.player);
  const { teamList } = useSelector((state: RootState) => state.team);

  // 既存のチーム情報取得
  const {
    isSuccess: getTeamSuccess,
    isError: getTeamError,
    refetch: getTeamRefetch,
  } = useGetTeamQuery();

  // チーム選択
  const handleSelectTeam = useCallback((myTeamId: string) => {
    dispatch(updateTeam(myTeamId));
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  // キャンセル
  const handleCancel = useCallback(() => {
    dispatch(updateTeam(''));
    navigate('/');
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return (
    <>
      <article className={styles.team}>
        <header className={styles.header}>
          <h1>Pick Your Team</h1>
        </header>
        <div>
          <ul className={styles.list}>
            {teamList.map(team => (
              <li key={team.id} className={styles.item}>
                <ListItem
                  id={team.id}
                  name={team.name}
                  disabled={getTeamError}
                  handleClick={() => {
                    handleSelectTeam(team.id);
                  }}
                />
              </li>
            ))}
          </ul>
          <AddListItem addClass={[styles.addForm]} />
        </div>
        <footer className={styles.footer}>
          <Button
            handleClick={() => {
              navigate('/play');
            }}
            disabled={!getTeamSuccess || !team}
          >
            START
          </Button>
          <Button handleClick={handleCancel}>cancel</Button>
        </footer>
      </article>
    </>
  );
};
