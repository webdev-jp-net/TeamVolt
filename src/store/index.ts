import { combineReducers, configureStore, getDefaultMiddleware } from '@reduxjs/toolkit';

import player, { playerGetApi, playerAddApi } from './player';
import team, { teamGetApi, teamPostApi } from './team';
// import user, { statusGetApi, statusUpdateApi } from './user';

const reducer = combineReducers({
  player,
  [playerGetApi.reducerPath]: playerGetApi.reducer,
  [playerAddApi.reducerPath]: playerAddApi.reducer,
  team,
  [teamGetApi.reducerPath]: teamGetApi.reducer,
  [teamPostApi.reducerPath]: teamPostApi.reducer,
  // user,
  // [statusGetApi.reducerPath]: statusGetApi.reducer,
  // [statusUpdateApi.reducerPath]: statusUpdateApi.reducer,
});

const middleware = getDefaultMiddleware({ serializableCheck: false }).concat(
  playerGetApi.middleware,
  playerAddApi.middleware,
  teamGetApi.middleware,
  teamPostApi.middleware
  // statusGetApi.middleware,
  // statusUpdateApi.middleware
);

export const store = configureStore({ reducer, middleware });

export type RootState = ReturnType<typeof reducer>;
