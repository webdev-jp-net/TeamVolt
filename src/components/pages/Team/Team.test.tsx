import { Router } from 'react-router-dom';

import { Provider } from 'react-redux';

import { render } from '@testing-library/react';
import { createMemoryHistory } from 'history';
import { store } from 'store';

import { Team } from './Team';
const history = createMemoryHistory();

describe('Team', () => {
  beforeEach(() => {
    window.scrollTo = jest.fn();
  });

  it('Teamのスナップショット', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <Router location={history.location} navigator={history}>
          <Team />
        </Router>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
