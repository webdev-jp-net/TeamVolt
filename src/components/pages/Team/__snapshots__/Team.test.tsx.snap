// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Team Teamのスナップショット 1`] = `
<DocumentFragment>
  <article
    class="team"
  >
    <header
      class="header"
    >
      <h1>
        Pick Your Team
      </h1>
    </header>
    <div>
      <ul
        class="list"
      />
      <div
        class="addListItem addForm"
      >
        <input
          class="input"
          placeholder="Your team not on the list?"
          type="text"
          value=""
        />
        <button
          class="button --basic"
          disabled=""
          type="button"
        >
          add
        </button>
      </div>
    </div>
    <footer
      class="footer"
    >
      <button
        class="button --basic"
        disabled=""
        type="button"
      >
        join
      </button>
      <button
        class="button --basic"
        type="button"
      >
        cancel
      </button>
    </footer>
  </article>
</DocumentFragment>
`;
